packages/permission-log-controller/src/PermissionLog.test.ts

```diff
@@ -0,0 +1,792 @@
+import { ControllerMessenger } from '@metamask/base-controller';
+import type {
+  JsonRpcEngineReturnHandler,
+  JsonRpcEngineNextCallback,
+} from '@metamask/json-rpc-engine';
+import type { PendingJsonRpcResponse } from '@metamask/utils';
+import { nanoid } from 'nanoid';
+import { useFakeTimers } from 'sinon';
+import type { SinonFakeTimers } from 'sinon';
+
+import { LOG_LIMIT, LOG_METHOD_TYPES } from './enums';
+import { PermissionLogController } from './PermissionLog';
+import type {
+  JsonRpcRequestWithOrigin,
+  Permission,
+  PermissionActivityLog,
+} from './PermissionLog';
+import { CaveatTypes, constants, getters, noop } from './permissions';
+
+const { PERMS, RPC_REQUESTS } = getters;
+const {
+  ACCOUNTS,
+  EXPECTED_HISTORIES,
+  SUBJECTS,
+  PERM_NAMES,
+  REQUEST_IDS,
+  RESTRICTED_METHODS,
+} = constants;
+
+let clock: SinonFakeTimers;
+
+type LogMiddleware = (
+  req: JsonRpcRequestWithOrigin,
+  res: PendingJsonRpcResponse<(string | Permission)[]>,
+  next?: JsonRpcEngineNextCallback,
+) => void;
```

Comment:
Hmm, I see why we have to use our own type. It might be good idea to see if we can reuse some pieces from `JsonRpcMiddleware` (we can think of this as a partial version). What if we do this instead:

``` suggestion
type LogMiddleware = JsonRpcMiddleware<JsonRpcParams, Json> extends (
  req: infer Request,
  res: infer Response,
  next: infer Next,
  end: infer End,
) => infer Result
  ? (req: Request, res: Response, next?: Next, end?: End) => Result
  : never;
```